### ISSUE TITLE
YOLOv3 architecture

### ISSUE THREAD

**Issue by jl749** [jl749/YOLOv3#1]
![image](https://user-images.githubusercontent.com/67103130/149910222-1419468b-4e1e-4039-ad38-a77767e96f70.png)
![image](https://user-images.githubusercontent.com/67103130/150452423-a8e587c8-fcb5-4d2d-9b89-baff2698d960.png)

YOLOv3 makes prediction across 3 different scales (13x13, 26x26, 52x52)<-- in case of 416x416 input.

The detection layer is used  to make prediction at feature maps of three different sizes, having strides 32, 16, 8 
```
416/32 = 13
416/16 = 26
416/8 = 52
```
In total predicts ((52 x 52) + (26 x 26) + 13 x 13)) x 3 = **10647 bounding boxes**

detection is done by using a 1x1 kernel on the feature maps 

Yolov3 uses independent logistic classifiers in place of the softmax function to determine the class of an input image. It also replaces the mean squared error with the binary cross-entropy loss, in simpler terms, the probability of object in the image and the class predictions are done using logistic regression.

[more read](https://blog.paperspace.com/how-to-implement-a-yolo-object-detector-in-pytorch/)

**Comment by jl749** [comment]
## What are the anchor boxes in YOLOv3?
After doing some clustering studies(K-means) on ground truth labels, it turns out that most bounding boxes have certain height-width ratios. So instead of directly predicting a bounding box, YOLOv2 (and v3) predict off-sets from a predetermined set of boxes with particular height-width ratios - those predetermined set of boxes are the anchor boxes.
Instead of predicting bounding boxes from scratch adjust clustered anchor boxes w, h. (easier + faster)
-darknet/issues/568-

![image](https://user-images.githubusercontent.com/67103130/150052753-e9df8a75-ed55-4cf1-a7fd-fc9d8d8ede50.png)
tx, ty, tw, th = logits (raw output of NN)
Pw, Ph = width, height of the pre clustered anchor boxes

sigmoid on tx, ty to put them in between 0 ~ 1

https://github.com/jl749/YOLOv3/blob/bc36f282c841d6c23940c1bb32d09f88d7c6aa86/yolov3/utils/functions.py#L112-L117

**Comment by jl749** [comment]
## torch.nn.Upsample()
![image](https://user-images.githubusercontent.com/67103130/150055834-c9b27cc3-d91d-4e26-b0c1-6d83db82fa74.png)

**Comment by jl749** [comment]
[YOLOv3 vs v4](https://ai-pool.com/a/s/yolov3-and-yolov4-in-object-detection)

**Comment by jl749** [comment]
# FPN (Feature Pyramid Network)
FPN is not an obj detector by itself. It is a feature extractor that works with object detector.

**GOAL**: feature integration

FPN is a general purpose architecture which is independent of the backbone

FPN consists of three components
- Bottom-up
- Top-down
- Lateral connection

![image](https://user-images.githubusercontent.com/67103130/150747866-fbfa9de3-a907-48d9-8c61-019835ab4fa1.png)

- YOLOv1 has used **(b)**
- YOLOv3 v4 v5 using **(d)**

Feature pyramids are a basic component in recognition systems for detecting objects at different scales.
It can detect different sized objects(small, medium, big) + less memory intensive


## problem with conventional way (b)
CNN computes a feature hierarchy layer by layer. However, there are **large semantic gaps** between layer

## solution using FPN
- Naturally leverage the pyramidal shape of a ConvNet's feature hierarchy while creating a feature pyramid that has strong semantics at all scales
- Combining semantically strong features with semantically weak features via a top-down pathway and lateral connection (d)
